File: code/backend/app/schemas/transaction.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel
from app.models.enums import EquipmentCondition, TransactionStatus


class IssueRequest(BaseModel):
    booking_id: int
    equipment_id: int


class ReturnRequest(BaseModel):
    condition_on_return: EquipmentCondition
    damage_report: Optional[str] = None


class TransactionRead(BaseModel):
    id: int
    booking_id: Optional[int]
    student_id: int
    equipment_id: int
    equipment_name: Optional[str] = None
    issued_at: datetime
    due_at: datetime
    returned_at: Optional[datetime]
    status: TransactionStatus
    condition_on_return: Optional[EquipmentCondition]
    damage_report: Optional[str]
    issued_by: int
    returned_to: Optional[int]

    model_config = {"from_attributes": True}

File: code/web_dashboard/src/types/transaction.ts
import type { EquipmentCondition } from './equipment';

export type TransactionStatus = 'ISSUED' | 'RETURNED' | 'RETURNED_DAMAGED' | 'OVERDUE';
export type DefaulterStatus = 'ACTIVE' | 'NOTIFIED' | 'RESOLVED';

export interface Transaction {
  id: number;
  booking_id?: number;
  student_id: number;
  equipment_id: number;
  equipment_name?: string;
  issued_by: number;
  issued_at: string;
  due_at: string;
  returned_at?: string;
  condition_on_return?: EquipmentCondition;
  damage_report?: string;
  status: TransactionStatus;
  returned_to?: number;
}

export interface Defaulter {
  id: number;
  student_id: number;
  student_name?: string;
  transaction_id: number;
  overdue_days: number;
  status: DefaulterStatus;
  detected_at: string;
  resolved_at?: string;
}
